subsys/net/lib/lwm2m/lwm2m_registry.c

```diff
@@ -1344,3 +1355,255 @@ bool lwm2m_engine_shall_report_obj_version(const struct lwm2m_engine_obj *obj)
 
 	return obj->version_major != 1 || obj->version_minor != 0;
 }
+
+#if defined(CONFIG_LWM2M_RESOURCE_DATA_CACHE_SUPPORT)
+static sys_slist_t lwm2m_timed_cache_list;
+static struct lwm2m_time_series_resource
+	lwm2m_cache_entries[CONFIG_LWM2M_RESOURCE_DATA_CACHE_ELEMENT_SIZE];
+
+static void lwm2m_cache_add_path_to_list(struct lwm2m_time_series_resource *new_entry)
+{
+	struct lwm2m_time_series_resource *prev = NULL;
+	struct lwm2m_time_series_resource *entry;
+
+	if (!sys_slist_is_empty(&lwm2m_timed_cache_list)) {
+
+		/* Keep list Alphabetical order */
+		SYS_SLIST_FOR_EACH_CONTAINER (&lwm2m_timed_cache_list, entry, node) {
+
+			if (strcmp(entry->path, new_entry->path) < 0) {
+				/* Current entry have  */
+				prev = entry;
+				continue;
+			}
+
+			if (prev) {
+				sys_slist_insert(&lwm2m_timed_cache_list, &prev->node, &new_entry->node);
+			} else {
+				sys_slist_prepend(&lwm2m_timed_cache_list, &new_entry->node);
+			}
+			return;
+		}
+	}
+
+	/* Add First or new tail entry */
+	sys_slist_append(&lwm2m_timed_cache_list, &new_entry->node);
+}
+
+static struct lwm2m_time_series_resource *lwm2m_cache_entry_allocate(char const *resource_path)
+{
+	int i;
+	struct lwm2m_time_series_resource *entry;
+
+	entry = lwm2m_cache_entry_get_by_string(resource_path);
+	if (entry) {
+		return entry;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(lwm2m_cache_entries); i++) {
+		if (lwm2m_cache_entries[i].path == NULL) {
+			lwm2m_cache_entries[i].path = resource_path;
+			lwm2m_cache_add_path_to_list(&lwm2m_cache_entries[i]);
+			return &lwm2m_cache_entries[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void lwm2m_engine_cache_write(struct lwm2m_engine_obj_field *obj_field, const char *pathstr,
+				     void *value)
+{
+	struct lwm2m_time_series_resource *cache_entry;
+	struct lwm2m_time_series_elem *elements_ptr;
+	bool updated_value = true;
+
+	cache_entry = lwm2m_cache_entry_get_by_string(pathstr);
+	if (!cache_entry) {
+		return;
+	} else if (cache_entry->size == cache_entry->capacity) {
+		LOG_WRN("Data cache full");
+		return;
+	}
+
+	elements_ptr = cache_entry->elements + cache_entry->size;
+
+	switch (obj_field->data_type) {
+	case LWM2M_RES_TYPE_U32:
+	case LWM2M_RES_TYPE_TIME:
+		elements_ptr->u32 = *(uint32_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_U16:
+		elements_ptr->u16 = *(uint16_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_U8:
+		elements_ptr->u8 = *(uint8_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_S64:
+		elements_ptr->i64 = *(int64_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_S32:
+		elements_ptr->i32 = *(int32_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_S16:
+		elements_ptr->i16 = *(int16_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_S8:
+		elements_ptr->i8 = *(int8_t *)value;
+		break;
+
+	case LWM2M_RES_TYPE_BOOL:
+		elements_ptr->b = *(bool *)value;
+		break;
+
+	case LWM2M_RES_TYPE_FLOAT:
+		elements_ptr->f = *(double *)value;
+		break;
+	default:
+		updated_value = false;
+		break;
+	}
+
+	if (updated_value) {
+		int rc;
```

Comment:
My understanding is that defining variables in the middle of a function is frowned upon.

Reply:
OK let me fix this.

Reply:
No, this is not wrong. C99 allows it and limiting the visibility/lifetime of a variable is always a good thing.